librz/core/linux_heap_glibc.c

```diff
@@ -1494,544 +2061,336 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
- * \param main_arena Pointer to Malloc state struct for main arena
- * \param global_max_fast The largest fast bin size (used for formatting)
- * \param format Enum to determine which type of bins to print.
- */
-static void GH(print_main_arena_bins)(RzCore *core, GHT m_arena, MallocState *main_arena, GHT global_max_fast, RzHeapBinType format) {
-	rz_return_if_fail(core && core->dbg && core->dbg->maps);
-	if (format == RZ_HEAP_BIN_ANY || format == RZ_HEAP_BIN_TCACHE) {
-		bool main_thread_only = true;
-		GH(print_tcache_instance)
-		(core, m_arena, main_arena, main_thread_only);
-		rz_cons_newline();
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	static GHT m_arena = GHT_MAX, m_state = GHT_MAX;
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	GHT global_max_fast = (64 * SZ / 4);
+	MallocState *main_arena = RZ_NEW0(MallocState);
+	if (!main_arena) {
+		return RZ_CMD_STATUS_ERROR;
 	}
-	if (format == RZ_HEAP_BIN_ANY || format == RZ_HEAP_BIN_FAST) {
-		char *input = malloc(sizeof(char) * 1);
-		input[0] = '\0';
-		bool main_arena_only = true;
-		GH(print_heap_fastbin)
-		(core, m_arena, main_arena, global_max_fast, input, main_arena_only);
-		free(input);
-		rz_cons_newline();
+	if (!GH(rz_resolve_main_arena)(core, &m_arena)) {
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
 	}
-	if (format == RZ_HEAP_BIN_ANY || format == RZ_HEAP_BIN_UNSORTED) {
-		GH(print_unsortedbin_description)
-		(core, m_arena, main_arena);
-		rz_cons_newline();
+	if (core->offset != core->prompt_offset) {
+		m_state = core->offset;
+	} else {
+		m_state = m_arena;
 	}
-	if (format == RZ_HEAP_BIN_ANY || format == RZ_HEAP_BIN_SMALL) {
-		GH(print_smallbin_description)
-		(core, m_arena, main_arena);
-		rz_cons_newline();
+	if (!GH(is_arena)(core, m_arena, m_state)) {
+		PRINT_RA("This address is not part of the arenas\n");
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	if (!GH(rz_update_main_arena)(core, m_state, main_arena)) {
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
+	}
+
+	bool json = false;
+	if (mode == RZ_OUTPUT_MODE_JSON) { // dmhdj
+		json = true;
+	}
+	RzHeapBinType bin_format = RZ_HEAP_BIN_ANY;
+	if (argc == 2) {
+		const char *input = argv[1];
+		if (!strcmp(input, "tcache")) {
+			bin_format = RZ_HEAP_BIN_TCACHE;
+		} else if (!strcmp(input, "fast")) {
+			bin_format = RZ_HEAP_BIN_FAST;
+		} else if (!strcmp(input, "unsorted")) {
+			bin_format = RZ_HEAP_BIN_UNSORTED;
+		} else if (!strcmp(input, "small")) {
+			bin_format = RZ_HEAP_BIN_SMALL;
+		} else if (!strcmp(input, "large")) {
+			bin_format = RZ_HEAP_BIN_LARGE;
+		}
+	}
+	GH(print_main_arena_bins)
+	(core, m_state, main_arena, global_max_fast, bin_format, json);
+	free(main_arena);
+	return RZ_CMD_STATUS_OK;
+}
+#ifndef HEAP32
+RZ_IPI RzCmdStatus rz_cmd_heap_chunks_print_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
+	if (core->rasm->bits == 64) {
```

Comment:
i'd suggest to define a small macro to avoid all these `if (core->rasm->bits == 64) ...._64 else ..._32`.